src/robot/libraries/OperatingSystem.py

```diff
@@ -286,34 +287,36 @@ def grep_file(self, path, pattern, encoding='UTF-8', encoding_errors='strict'):
         Possible trailing newline is never returned.
 
         A line matches if it contains the ``pattern`` anywhere in it and
-        it *does not need to match the pattern fully*. The pattern
-        matching syntax is explained in `introduction`, and in this
+        it *does not need to match the pattern fully when not in regex mode*. 
+        The pattern matching syntax is explained in `introduction`, and in this
         case matching is case-sensitive.
 
         Examples:
         | ${errors} = | Grep File | /var/log/myapp.log | ERROR |
         | ${ret} = | Grep File | ${CURDIR}/file.txt | [Ww]ildc??d ex*ple |
+        | ${ret} = | Grep File | ${CURDIR}/file.txt | [Ww]ildc..d ex.*ple | regexp=True
 
-        If more complex pattern matching is needed, it is possible to use
-        `Get File` in combination with String library keywords like `Get
-        Lines Matching Regexp`.
+        If more complex pattern matching is needed, it is possible to set
+        the regexp flag to True. When in regex mode, the pattern need to match fully.
 
         This keyword supports special ``SYSTEM`` and ``CONSOLE`` encodings that
         `Get File` supports only with Robot Framework 4.0 and newer. When using
         Python 3, it is possible to use ``${NONE}`` instead of ``SYSTEM`` with
         earlier versions.
         """
         path = self._absnorm(path)
-        pattern = '*%s*' % pattern
+        if is_falsy(regexp):
+            pattern = fnmatch.translate('*%s*' % pattern)
+        reobj = re.compile(pattern)
```

Comment:
The * around the pattern for the re.match pattern is still needed, otherwise it does not trigger when switching to search.
The hint about re.search, was nice though! 
The comment about is_falsy noted.

Reply:
The reason is because `fnmatch.translate` adds a \Z to the regular expression, matching only end of string.

`>>> import fnmatch, re
>>>
>>> regex = fnmatch.translate('*.txt')
>>> regex
'(?s:.*\\.txt)**\\Z**'
`